Refactor/atomic design pattern

geon-2 opened the pull request:
## ✨ Issue Number

- #9 #10 

## 🗂️ Details

- Atoms: 최하위 컴포넌트
- [ ] Button
- [ ] Text
- [ ] Wrapper
- Mocules: atom으로 구성된 컴포넌트
- [ ] Item
- [ ] CalendarItem
- [ ] StatusTag

- Button Components
- StarToggleButton: 즐겨찾기 버튼
- MainButton: submit 버튼

- Text Components
- Text: 기본 텍스트 (size, weight 등 커스텀 가능함)
- ThemeText: 테마 고유 텍스트 (자주 사용되는 텍스트가 지정됨)

- Mocules
- Item: 공고 리스트 아이템
- CalendarItem: 공고 캘린더 아이템 (날짜)
- StatusTag: 상태 표시 태그

- 'http://localhost:5173/theme-preview'으로 preview 확인 가능

## ✅ Need Review
- 삭제 가능한 코드의 여부
- 모듈로서의 적합성
- 추가적으로 필요한 기능 여부

## 📸 Screenshot
![Vite-React-TS-12-10-2024_11_07_PM](https://github.com/user-attachments/assets/67bf6ff3-474e-46de-9971-e1a003684ef2)

## 📚 Reference
[Atomic Design Pattern](https://velog.io/@blee94/React-Atomic-Pattern-%EC%95%84%ED%86%A0%EB%AF%B9-%EB%94%94%EC%9E%90%EC%9D%B8%ED%8C%A8%ED%84%B4-%EC%97%90-%EB%8C%80%ED%95%B4-ARABOZA)

geon-2 committed:
Feat: Add message communication feature
메시지 통신 기능 추가
- /global.d.ts
    - window.ReactNativeWebView 타입 오류 방지

- api 수정 ( /src/api )
    - /message
        - 메시지 데이터 타입 ( MessageType )
        - 메시지 전송 메소드 추가 ( sendMessage )
    - /interface
        - 보조출연자 촬영 관리용 데이터 인터페이스 정의 ( RecruitStatus, Recruit, RecruitList )
    - /dummyData
        - 보조출연자 촬영 관리용 더미데이터 추가 ( dummyRecruitList )

- redux 수정
    - 웹뷰로 띄워졌을 때 전역 상태가 초기화 되는 문제가 발생함
    - redux-persist로 localStorage에 전역 상태를 저장하여 유지하는 방식으로 해결
    - /src/main.tsx
        - redux-persist 설정을 위한 PersistGate 추가
    - /redux/rootReducer
        - slice별 config 상수 초기화
        - persistReducer를 통해 각 슬라이스 별 리듀서 초기화
        - combineReducers를 통해 모든 리듀서를 rootReducer로 통합
    - /redux/store
        - rootReducer 호출
        - 미들웨어 설정
        - persistor 추출

- /src/pages & /src/components
    - 사용하지 않는 코드 주석화
    - navigate => sendMessage
    - ExtraShootManagePage, DropDownSelector 추가
        - 테스트 목적으로 feature/extra-shoot-page 브랜치 내의 코드에서 복사
    - DetailPage 명칭 변경 => CompanyDetailPage
    - StatusRecruitBox 수정
        - swiped state가 전역으로 설정되어 있어서 모든 요소들이 해당 상태롤 공유하는 문제가 발생
        - 컴포넌트의 고유한 state로 관리할 수 있도록 설정
        - 클릭 시 RN 라우팅이 실행되도록 메시지 전송 로직 추가
        - 전역 상태로 관리되던 공고 상태 값들은 추가한 더미 데이터리스트를 사용하여 각 공고 데이터에서 받아오도록 수정
    - NavBar 수정
        - back 버튼 클릭 시 RN 상에서 뒤로 가기가 작동하도록 수정
    - CompleteModal 수정
        - 홈 화면, 촬영 관리로 라우팅 되도록 수정

geon-2 committed:
Feat: Add admin manage page

geon-2 committed:
Merge branch 'main' into feature/message-communication

geon-2 committed:
Feat: Add utils
api 유틸리티 관련 유틸리티 추가
- sendMessage: 메시지 통신 (message.ts에서 이동)
- decryptedAccessToken: 토큰 복호화
- verifyHmacSignature: 토큰 서명 확인
- requestToken: 토큰 요청 (localStorage에 없으면 RN에 요청)
- requestFetch: fetch API 요청
- requestPostFetch: POST 요청
- requestPutFetch: PUT 요청
- requestGetFetch: GET 요청
- requestDeleteFetch: DELETE 요청

geon-2 committed:
Merge branch 'develop' into feature/message-communication

geon-2 committed:
Fix: Update file path
인식되지 않는 파일 경로 수정

geon-2 committed:
Feat: Update CompanyDateSelectedPage

geon-2 committed:
Feat: Update the pages related to company

geon-2 committed:
Merge branch 'develop' into feature/message-communication

geon-2 committed:
Chore: Install crypto-js

geon-2 committed:
Feat: Add ApplicantDetail page

geon-2 committed:
Merge branch 'feature/real-manager-page' of https://github.com/UMC-EXTRA/extra-react-native-webview into feature/manager-page

geon-2 committed:
Merge branch 'feature/manager-page' into feature/message-communication

geon-2 committed:
Feat: Fix add notice page

geon-2 committed:
Merge branch 'develop' into feature/message-communication

geon-2 committed:
Feat: Connect notice register page to server

geon-2 committed:
Fix: StatusRecruitBox routing

geon-2 committed:
Merge branch 'develop' into feature/message-communication

geon-2 committed:
Feat: Connect RN

geon-2 committed:
Fix: Fix build error in jobPostAPIForCom.ts

geon-2 committed:
Merge branch 'develop' into feature/message-communication

geon-2 committed:
Fix: build error

geon-2 committed:
Fix: build error

geon-2 committed:
Fix: build error

geon-2 committed:
Fix: package error

geon-2 committed:
Chore: Add vercel.json

geon-2 committed:
Fix: routing error

geon-2 committed:
Fix: manage page routing

geon-2 committed:
Fix: ExtraCastingBoard routing error

geon-2 committed:
Fix: manage page routing error

geon-2 committed:
Fix: click event error

geon-2 committed:
Fix: Change message type

geon-2 committed:
Refactor: Atomic Design Pattern 추가
atoms, mocules 추가
- atoms: 최하위 구성 컴포넌트
    - Text
    - Button
    - Wrapper: atoms를 묶는 컨테이너 역할
- mocules: atoms로 구성된 컴포넌트
    - Item: 공고 리스트 아이템
    - CalendarItem: 캘린터 아이템

TODO
- Form 관련 컴포넌트 구현
    - InputField
    - Select
    - Checkbox
    - DropDownList
- Organisms 추가: mocules로 구성된 컴포넌트
    - Header
    - BottomNavigation
    - Calendar
    - RecruitList
- Theme 기능 추가 (미정)
    - light/dark theme

geon-2 committed:
Refactor: 상태 태그 및 주석 추가
- mocules에 상태 태그 추가
    - StatusTag
- 가독성을 위한 주석 추가

nyewon commented on src/components/atoms/Button.tsx:
```diff
@@ -0,0 +1,57 @@
+  background: ${({ isActive }) => (isActive ? "#f5c001" : "#575757")};
+`;
+
+const StarToggleButton: React.FC<ButtonProps> = ({ onClick, isActive }) => {
```
함수형 컴포넌트에서 타입을 지정할 때 React.FC의 사용은 지양해야 한다고 알고 있습니다!

React.FC를 사용할 경우 기본적으로 children props가 포함되어 props 타입이 명확하지 않다는 단점이 있습니다. 컴포넌트에서 children을 사용하지 않는 경우에도 불필요하게 children이 props 타입에 포함되기 때문에 코드의 의도와 타입 정의가 불일치할 수 있습니다.

따라서 React.FC를 사용하는 것 보다
`const StarToggleButton = ({ onClick, isActive }: ButtonProps)`
이런식으로 props를 명시적으로 지정하는게 더 좋을 것 같습니다!

참고: https://velog.io/@yena1025/FunctionComponent-FC-%EC%82%AC%EC%9A%A9-%EC%A4%84%EC%9D%B4%EA%B8%B0-24jhm0wp

geon-2 replied:
오 전혀 몰랐던 내용이네요. 알려주셔서 감사합니다! 이거는 바로 수정해서 올릴께요

goodbyebada commented on src/components/mocules/CalendarItem.tsx:
```diff
@@ -0,0 +1,67 @@
+ * @param onClick () => void (onClick method)
+ * @returns
+ */
+const CalendarItem: React.FC<CalendarItemProps> = ({
+  children,
+  count = 0,
+  highlight = false,
+  onClick = () => {},
+}) => {
+  return (
+    <StyledCalendarItem onClick={onClick}>
+      {count > 0 && (
+        <CounterWrapper>
+          <Text size={16} weight={900}>
```
현재 CalendarItem 컴포넌트의 글자 크기가 props로 주입받는 방식이 아니라, 내부에서 고정값 16으로 설정되어 있네요!
화면 크기에 따라 글자 크기도 유동적으로 변경될 수 있도록, props로 size를 주입받는 방식은 어떨까요?

또한, 만약 props로 글자 크기를 전달받게 된다면, 아래와 같이 font size를 상수로 관리하는 방식을 추천합니다! fontsize를 통일하기도 쉽고 유지보수성이 높아질 것 같습니다
```js
const FONT_SIZE = {
  LARGE: 30,
  MEDIUM: 16,
  SMALL: 10,
};

// 사용 예시
<Text size={FONT_SIZE.SMALL} weight={900} />
```

geon-2 replied:
일단 숫자로 지정가능하도록 변경했어요. fontsize 통일하는 작업은 테마 작업하면서 추가할께요!

goodbyebada commented on src/components/mocules/Item.tsx:
```diff
@@ -0,0 +1,123 @@
+  }
+`;
+
+/**
+ * Recruitment List Item
+ * highlight: 추천 공고 시 지정
+ * isActiveStar: 즐겨찾기 활성화 여부
+ * isRecruit: 즐겨찾기 버튼 표시 여부
+ * @param title string (title)
+ * @param category string (category)
+ * @param date string (date: 01/01 - 01/02)
+ * @param dDay string (D-Day)
+ * @param company string (company name)
+ * @param time string (time: 01:00)
+ * @param location string (location)
+ * @param status "applied" | "rejected" | "approved" (status)
+ * @param statusText string (status text)
+ * @param highlight boolean (specify when posting a recommendation)
+ * @param isActiveStar boolean (favorites status)
+ * @param isRecruit boolean (whether to enable the favorites button)
+ */
```
주석이 자세히 작성되어 있어 이해하기 쉽고, 다른 사람이 코드를 참고하거나 수정할 때 큰 도움이 될 것 같습니다! 👍

goodbyebada commented on src/components/atoms/Text.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  ${({ inline }) => (inline ? `display: inline-block;` : "")}
+
+  color: ${({ color, highlight }) => (highlight ? "#F5C001" : color || "#fff")};
```
현재 색깔이 문자열로 지정되어있습니다. 색깔도 상수화하는 것은 어떤가요?

```js
const COLORS = {
	HIGH_LIGHT : "#F5C001"
}

//사용시 
color: ${({ color, highlight }) => (highlight ? COLORS.HIGH_LIGHT : color || "#fff")};
```

이렇게 하면 디자인 변경으로 highlight 색상이 다른 색으로 바뀌어도 COLORS.HIGH_LIGHT의 값만 수정하면 일괄적으로 수정이 적용되는 장점이 있습니다.

저희끼리 highlight, main ,emphasis (강조)등 색과 명칭에 대해 합의 후 적용하는건 어떤가요?

geon-2 replied:
좋은 의견인 것 같네요! 상수화는 테마 작업할 때 통일 시키려고 했는데, 회의 때 정리해서 통일시켜보죠!

geon-2 committed:
Refactor: React.FC 구문 삭제
React.FC에 포함된 children props 때문에 타입이 명확하지 않다는 의견
React.FC 삭제 -> props의 명시적인 타입 지정

geon-2 committed:
Refactor: CalendarItem에 fontSize props 추가
CalendarItem의 글자크기가 고정 되어 있어, fontSize props를 통해 수정 가능하도록 변경

nyewon reviewed (approved):
변경된 내용 확인했습니다! 고생하셨습니다 :)

goodbyebada reviewed (approved):
확인했습니다. 👍
